File: app/src/main/java/com/mcal/disassembler/activities/SymbolsActivity.java
package com.mcal.disassembler.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.BaseAdapter;
import android.widget.ListView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;

import com.mcal.disassembler.R;
import com.mcal.disassembler.nativeapi.Dumper;
import com.mcal.disassembler.view.CenteredToolBar;
import com.mcal.disassembler.widgets.SnackBar;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class SymbolsActivity extends AppCompatActivity {
    SnackBar bar;
    private List<Map<String, Object>> data;
    private String path;
    private CenteredToolBar toolbar;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.symbols_activity);
        setupToolbar(getString(R.string.app_symbols));
        ListView list = findViewById(R.id.symbols_activity_list_view);
        data = getData();
        SymbolsAdapter adapter = new SymbolsAdapter(this);
        list.setAdapter(adapter);
        list.setOnItemClickListener(new ItemClickListener());

        path = Objects.requireNonNull(getIntent().getExtras()).getString("filePath");
    }

    @SuppressWarnings("ConstantConditions")
    private void setupToolbar(String title) {
        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle(title);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
    }

    @NotNull
    private List<Map<String, Object>> getData() {
        List<Map<String, Object>> list = new ArrayList<>();
        Map<String, Object> map;
        for (int i = 0; i < Dumper.symbols.size(); ++i) {
            map = new HashMap<>();
            if (Dumper.symbols.get(i).getType() == 1)
                map.put("img", R.drawable.ic_box_blue);
            else if (Dumper.symbols.get(i).getType() == 2)
                map.put("img", R.drawable.ic_box_red);
            else map.put("img", R.drawable.ic_box_green);
            map.put("title", Dumper.symbols.get(i).getDemangledName());
            map.put("info", Dumper.symbols.get(i).getName());
            map.put("type", Dumper.symbols.get(i).getType());
            list.add(map);
        }
        return list;
    }

    public void showSearch(View view) {
        Intent i = new Intent(this, SearchActivity.class);
        Bundle bundle = new Bundle();
        bundle.putString("filePath", path);
        i.putExtras(bundle);
        startActivity(i);
    }

    public void showMenu(View view) {
        Intent i = new Intent(this, MenuActivity.class);
        Bundle bundle = new Bundle();
        bundle.putString("filePath", path);
        i.putExtras(bundle);
        startActivity(i);
    }

    @Override
    public void onBackPressed() {
        bar = new SnackBar(this, getString(R.string.againToExit));
        bar.show();
        bar.setDismissTimer(2500);
        bar.setOnBackPressedListener(SymbolsActivity.this::finish);
    }

    @Override
    public boolean onOptionsItemSelected(@NotNull MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    static class ViewHolder {
        public AppCompatTextView info;
        public int type;
        AppCompatImageView img;
        AppCompatTextView title;
    }

    private final class ItemClickListener implements OnItemClickListener {
        @Override
        public void onItemClick(AdapterView<?> arg0, @NotNull View view, int arg2, long arg3) {
            Bundle bundle = new Bundle();
            bundle.putString("demangledName", (String) (((ViewHolder) view.getTag()).title.getText()));
            bundle.putString("name", (String) (((ViewHolder) view.getTag()).info.getText()));
            bundle.putInt("type", ((ViewHolder) view.getTag()).type);
            bundle.putString("filePath", path);
            Intent intent = new Intent(SymbolsActivity.this, SymbolActivity.class);
            intent.putExtras(bundle);
            startActivity(intent);
        }
    }

    public class SymbolsAdapter extends BaseAdapter {
        private LayoutInflater mInflater = null;

        private SymbolsAdapter(Context context) {
            this.mInflater = LayoutInflater.from(context);
        }

        @Override
        public int getCount() {
            return data.size();
        }

        @Override
        public Object getItem(int position) {
            return position;
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            ViewHolder holder;

            if (convertView == null) {
                holder = new ViewHolder();
                convertView = mInflater.inflate(R.layout.symbol_list_item, null);
                holder.img = convertView.findViewById(R.id.symbolslistitemimg);
                holder.title = convertView.findViewById(R.id.symbolslistitemTextViewtop);
                holder.info = convertView.findViewById(R.id.symbolslistitemTextViewbottom);
                convertView.setTag(holder);
            } else {
                holder = (ViewHolder) convertView.getTag();
            }
            holder.img.setBackgroundResource((Integer) data.get(position).get("img"));
            holder.title.setText((String) data.get(position).get("title"));
            holder.info.setText((String) data.get(position).get("info"));
            holder.type = ((int) data.get(position).get("type"));

            return convertView;
        }
    }
}

File: app/src/main/java/com/mcal/disassembler/activities/SymbolsActivity.kt
package com.mcal.disassembler.activities

import android.graphics.Color
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.mcal.disassembler.R
import com.mcal.disassembler.adapters.SymbolsItem
import com.mcal.disassembler.data.Preferences
import com.mcal.disassembler.data.Storage
import com.mcal.disassembler.databinding.ActivitySymbolsBinding
import com.mcal.disassembler.databinding.ProgressDialogBinding
import com.mcal.disassembler.nativeapi.Dumper
import com.mcal.disassembler.utils.FileHelper
import com.mcal.disassembler.view.FloatingButton
import com.mcal.disassembler.view.SnackBar
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.adapters.ItemAdapter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class SymbolsActivity : SymbolsSearchActivity() {
    private val binding by lazy(LazyThreadSafetyMode.NONE) {
        ActivitySymbolsBinding.inflate(layoutInflater)
    }
    private var dialogBinding: ProgressDialogBinding? = null

    private var mPath: String? = null
    private var dialog: AlertDialog? = null
    private var mBar: SnackBar? = null

    private val itemAdapter = ItemAdapter<SymbolsItem>()
    private val fastAdapter = FastAdapter.with(itemAdapter)

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        setupToolbar(binding.toolbar, getString(R.string.app_symbols))
        intent.extras?.let { bundle ->
            val path = bundle.getString("path").also { path ->
                mPath = path
            }
            if (path != null) {
                val recyclerView = binding.symbolsActivityListView.apply {
                    adapter = fastAdapter
                }

                if (data.isEmpty()) {
                    initData()
                }

                val searchText = binding.search
                val clearBtn = binding.clearText.apply {
                    setOnClickListener {
                        searchText.setText("")
                    }
                }
                searchText.addTextChangedListener(object : TextWatcher {
                    override fun onTextChanged(
                        s: CharSequence,
                        start: Int,
                        before: Int,
                        count: Int
                    ) = Unit

                    override fun beforeTextChanged(
                        s: CharSequence,
                        start: Int,
                        count: Int,
                        after: Int
                    ) = Unit

                    override fun afterTextChanged(s: Editable) {
                        setVisibility(clearBtn, if (s.isEmpty()) View.GONE else View.VISIBLE)
                        if (canStartFilterProcess) {
                            if (!TextUtils.equals(s, lastValue)) {
                                val constraint = s.toString()
                                lastValue = constraint
                                recyclerView.scrollToPosition(0)
                                canStartFilterProcess = false
                                filter(constraint)
                                return
                            }
                            return
                        }
                        newValue = s.toString()
                    }
                })
                binding.saveSymbols.setOnClickListener {
                    showProgressDialog()
                    mBar = SnackBar(this, getString(R.string.done)).also { snackBar ->
                        CoroutineScope(Dispatchers.IO).launch {
                            val size = Dumper.symbols.size
                            val symbols = arrayOfNulls<String>(size)
                            val demangledSymbols = arrayOfNulls<String>(size)
                            for (i in Dumper.symbols.indices) {
                                withContext(Dispatchers.Main) {
                                    updateDialogProgress(i, size)
                                }
                                symbols[i] = Dumper.symbols[i].name
                                demangledSymbols[i] = Dumper.symbols[i].demangledName
                            }
                            val storage = Storage(this@SymbolsActivity)
                            FileHelper.writeSymbolsToFile(
                                storage.getSymbolsDir(),
                                "Symbols.txt",
                                symbols
                            )
                            FileHelper.writeSymbolsToFile(
                                storage.getSymbolsDir(),
                                "Symbols_demangled.txt",
                                demangledSymbols
                            )
                            withContext(Dispatchers.Main) {
                                dismissProgressDialog()
                                snackBar.show()
                            }
                        }
                    }
                }
                binding.showFloatingMenu.setOnClickListener {
                    FloatingButton(this, path).show()
                }
                val preferences = Preferences(this)
                binding.regex.setBackgroundColor(
                    if (preferences.regex) ActivityCompat.getColor(
                        this,
                        R.color.colorAccent
                    ) else {
                        Color.TRANSPARENT
                    }
                )
                binding.regex.setOnClickListener {
                    if (preferences.regex) {
                        preferences.regex = false
                        binding.regex.setBackgroundColor(Color.TRANSPARENT)
                    } else {
                        preferences.regex = true
                        binding.regex.setBackgroundColor(
                            ActivityCompat.getColor(
                                this,
                                R.color.colorAccent
                            )
                        )
                    }
                }
            }
        }
    }

    private fun initData() {
        setVisibility(binding.progress, View.VISIBLE)
        val list = symbolsFilteredList
        if (list.isNotEmpty()) {
            list.clear()
        }
        var map: Map<String, Any>
        for (i in Dumper.symbols.indices) {
            map = HashMap()
            when (Dumper.symbols[i].type) {
                1 -> map["img"] = R.drawable.ic_box_blue
                2 -> map["img"] = R.drawable.ic_box_red
                else -> map["img"] = R.drawable.ic_box_green
            }
            map["title"] = Dumper.symbols[i].demangledName
            map["info"] = Dumper.symbols[i].name
            map["type"] = Dumper.symbols[i].type
            list.add(map)
        }
        list.sortBy {
            it["title"] as String
        }
        updateSymbolsSize(list)
        updateAdapter(list)
        val dataList = data
        if (dataList.isNotEmpty()) {
            dataList.clear()
        }
        dataList.addAll(list)
    }

    private fun updateSymbolsSize(list: ArrayList<Map<String, Any>>) {
        val symbolsSizeView = binding.symbolsSize
        val dataSize = list.size.toString()
        if (symbolsSizeView.text.toString() != dataSize) {
            symbolsSizeView.text = buildString {
                append(getString(R.string.symbols_count))
                append(dataSize)
            }
        }
    }

    private fun showProgressDialog() {
        dialog = MaterialAlertDialogBuilder(this).apply {
            dialogBinding = ProgressDialogBinding.inflate(layoutInflater).also { binding ->
                setView(binding.root)
            }
            setCancelable(false)
            setTitle(R.string.saving)
        }.create().also {
            it.show()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        dismissProgressDialog()
    }

    private fun dismissProgressDialog() {
        dialog?.takeIf { it.isShowing }?.let {
            it.dismiss()
            dialog = null
            dialogBinding = null
        }
    }

    private fun updateDialogProgress(last: Int, total: Int) {
        dialogBinding?.let { binding ->
            binding.progress.apply {
                progress = last
                max = total
            }
            binding.count.text = buildString {
                append(last)
                append(" / ")
                append(total)
            }
        }
    }

    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        SnackBar(this, getString(R.string.againToExit)).apply {
            show()
            dismissTimer = 2500
            setOnBackPressedListener { finish() }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun updateAdapter(list: ArrayList<Map<String, Any>>) {
        val path = mPath
        if (path != null) {
            val adapter = itemAdapter
            if (adapter.adapterItemCount >= 0) {
                adapter.clear()
            }
            var item: Map<String, Any>
            for (i in list.indices) {
                item = list[i]
                adapter.add(
                    SymbolsItem()
                        .withContext(this)
                        .withId(i.toLong())
                        .withIcon(item["img"] as Int)
                        .withTitle(item["title"] as String)
                        .withSubTitle(item["info"] as String)
                        .withSymbolType(item["type"] as Int)
                        .withPath(path)
                )
            }
        }
        setVisibility(binding.progress, View.GONE)
    }

    override fun onStartSearch() {
        setVisibility(binding.progress, View.VISIBLE)
    }

    override fun onFoundApp(list: ArrayList<Map<String, Any>>, mode: Boolean) {
        setVisibility(
            binding.symbolsNotFound, if (mode) {
                View.GONE
            } else {
                View.VISIBLE
            }
        )
        setVisibility(
            binding.symbolsActivityListView, if (mode) {
                View.VISIBLE
            } else {
                View.GONE
            }
        )
        updateSymbolsSize(list)
        updateAdapter(list)
    }
}